File: internal/targets/matcher.go
package targets

type Pattern struct {
	Exclude []string
	Include []string
}

func MatchAny(patterns []string, path string) (bool, error) {
	for _, pattern := range patterns {
		matched, err := GitignoreMatch(pattern, path)
		if err != nil {
			return false, err
		}
		if matched {
			return true, nil
		}
	}
	return false, nil
}

// Is ignored for `exclude` or `include` or `fallback`.
func Ignores(internal, external Pattern, ctx *MatcherContext, name string, def bool) bool {
	check := false
	var err error

	check, err = MatchAny(internal.Exclude, name)
	if err != nil {
		goto Error
	}
	if check {
		return true
	}

	check, err = MatchAny(internal.Include, name)
	if err != nil {
		goto Error
	}
	if check {
		return false
	}

	check, err = MatchAny(external.Exclude, name)
	if err != nil {
		goto Error
	}
	if check {
		return true
	}

	check, err = MatchAny(external.Include, name)
	if err != nil {
		goto Error
	}
	if check {
		return false
	}

	return def

Error:
	ctx.SourceErrors = append(ctx.SourceErrors, err)
	return false
}

File: src/matcher.ts
import { gitignoreMatch } from './gitignore.js'
import type { Matcher, TargetContext } from './scan.js'

function matchAny(patterns: string[], path: string): boolean {
  for (const pattern of patterns) {
    const matched = gitignoreMatch(pattern, path)
    if (matched) {
      return true
    }
  }
  return false
}

export function ignores(matcher: Matcher, ctx: TargetContext, name: string, def: boolean): boolean {
  let check = false

  try {
    check = matchAny(matcher.internal.exclude, name)
    if (check) {
      return true
    }

    check = matchAny(matcher.internal.include, name)
    if (check) {
      return false
    }

    check = matchAny(matcher.external.exclude, name)
    if (check) {
      return true
    }

    check = matchAny(matcher.external.include, name)
    if (check) {
      return false
    }
  }
  catch (err) {
    ctx.sourceErrors.push(err as Error)
    return false
  }

  return def
}
